ath79: add support for MikroTik hAP (RB951Ui-2nD)

mkg20001 opened the pull request:
The MikroTik hAP ac lite (product code RB951Ui-2nD) is
an indoor 2.4Ghz AP with a 2 dBi integrated antenna built around the
Atheros QCA9531 SoC.

Specifications:
 - SoC: Atheros QCA9531
 - RAM: 64 MB
 - Storage: 16 MB NOR - Winbond 25Q128FVSG
 - Wireless: Atheros QCA9530 (SoC) 802.11b/g/n 2x2
 - Ethernet: Atheros AR934X switch, 5x 10/100 ports,
   10-28 V passive PoE in port 1, 500 mA PoE out on port 5
 - 8 user-controllable LEDs:
   · 1x power (green)
   · 1x user (green)
   · 4x LAN status (green)
   · 1x WAN status (green)
   · 1x PoE power status (red)

See mikrotik.com/product/RB951Ui-2nD for more details.

Notes:
 The device was already supported in the ar71xx target.

Flashing:
 TFTP boot initramfs image and then perform sysupgrade. Follow common
 MikroTik procedure as in openwrt.org/toh/mikrotik/common.

Signed-off-by: Maciej Krüger <mkg20001@gmail.com>

mkg20001 commented:
I'm not sure if the partitions are correct (the most important stufff seems correct)

Also switch config is a bit different, it's now eth0.1@eth0 instead of eth1.1@eth1. Not sure if that's problematic.

This depends on #3348 (it's mostly a copy with some adjustments)

mtd

```
19.07

dev:    size   erasesize  name
mtd0: 0000e000 00010000 "routerboot" !!
mtd1: 00001000 00010000 "hard_config" !!
mtd2: 00001000 00010000 "bios"
mtd3: 0000f000 00010000 "routerboot2"
mtd4: 00001000 00001000 "soft_config"
mtd5: 00fe0000 00010000 "firmware"
mtd6: 001b0000 00010000 "kernel"
mtd7: 00e30000 00010000 "rootfs"
mtd8: 00bc0000 00010000 "rootfs_data"

current port

dev:    size   erasesize  name
mtd0: 00020000 00010000 "RouterBoot" !!
mtd1: 0000e000 00010000 "bootloader1" !!
mtd2: 00001000 00010000 "hard_config" !!
mtd3: 00001000 00010000 "bios"
mtd4: 0000f000 00010000 "bootloader2"
mtd5: 00001000 00010000 "soft_config"
mtd6: 00fe0000 00010000 "firmware"
mtd7: 001b0000 00010000 "kernel"
mtd8: 00e30000 00010000 "rootfs"
mtd9: 00bc0000 00010000 "rootfs_data"
```

```
19.07 ip a
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host 
       valid_lft forever preferred_lft forever
2: eth0: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc fq_codel state DOWN qlen 1000
    link/ether b8:69:f4:0d:1a:36 brd ff:ff:ff:ff:ff:ff
3: eth1: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc fq_codel state UP qlen 1000
    link/ether b8:69:f4:0d:1a:37 brd ff:ff:ff:ff:ff:ff
    inet6 fe80::ba69:f4ff:fe0d:1a37/64 scope link 
       valid_lft forever preferred_lft forever
4: wlan0: <BROADCAST,MULTICAST> mtu 1500 qdisc noop state DOWN qlen 1000
    link/ether b8:69:f4:0d:1a:3b brd ff:ff:ff:ff:ff:ff
5: br-lan: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP qlen 1000
    link/ether b8:69:f4:0d:1a:37 brd ff:ff:ff:ff:ff:ff
    inet 192.168.1.1/24 brd 192.168.1.255 scope global br-lan
       valid_lft forever preferred_lft forever
    inet6 fdc2:b285:9bb3::1/60 scope global 
       valid_lft forever preferred_lft forever
    inet6 fe80::ba69:f4ff:fe0d:1a37/64 scope link 
       valid_lft forever preferred_lft forever
6: eth1.1@eth1: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue master br-lan state UP qlen 1000
    link/ether b8:69:f4:0d:1a:37 brd ff:ff:ff:ff:ff:ff

current ip a
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host 
       valid_lft forever preferred_lft forever
2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc fq_codel state UP qlen 1000
    link/ether e2:3e:65:92:0b:bb brd ff:ff:ff:ff:ff:ff
    inet6 fe80::e03e:65ff:fe92:bbb/64 scope link 
       valid_lft forever preferred_lft forever
3: eth1: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc fq_codel state DOWN qlen 1000
    link/ether b8:69:f4:0d:1a:36 brd ff:ff:ff:ff:ff:ff
5: wlan0: <BROADCAST,MULTICAST> mtu 1500 qdisc noop state DOWN qlen 1000
    link/ether b8:69:f4:0d:1a:3c brd ff:ff:ff:ff:ff:ff
6: br-lan: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP qlen 1000
    link/ether b8:69:f4:0d:1a:37 brd ff:ff:ff:ff:ff:ff
    inet 192.168.1.1/24 brd 192.168.1.255 scope global br-lan
       valid_lft forever preferred_lft forever
    inet6 fdb7:4e2a:a0bb::1/60 scope global noprefixroute 
       valid_lft forever preferred_lft forever
    inet6 fe80::ba69:f4ff:fe0d:1a37/64 scope link 
       valid_lft forever preferred_lft forever
7: eth0.1@eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue master br-lan state UP qlen 1000
    link/ether b8:69:f4:0d:1a:37 brd ff:ff:ff:ff:ff:ff
```

joaochainho commented:
Hi @mkg20001 , many thanks for your work on supporting this device! Have a couple lying around, so I'm happy to test it 👍 
Just a small detail: the title refers to hAP ac, but I afaict this isn't the ac version, just hAP.

mkg20001 commented:
Removed ath10k, wifi still works. I have trouble getting it to boot with kmod-usb2. Since I don't really want to mess arround with it and stick a serial console adapter in there, is there anything else I could use to debug?

f00b4r0 commented:
> I'm not sure if the partitions are correct (the most important stufff seems correct)
> 
> Also switch config is a bit different, it's now eth0.1@eth0 instead of eth1.1@eth1. Not sure if that's problematic.
> 
> This depends on #3348 (it's mostly a copy with some adjustments)

Having two devices in the same PR complicates review. You should push to this PR only the changes related to this device.

Git can merge the two separate patchsets.

> mtd
> 
> ```
> mtd0: 00020000 00010000 "RouterBoot" !!
> mtd1: 0000e000 00010000 "bootloader1" !!
> mtd2: 00001000 00010000 "hard_config" !!
> ```

That's ok.

>  ```
> 19.07 ip a
> 2: eth0: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc fq_codel state DOWN qlen 1000
>     link/ether b8:69:f4:0d:1a:36 brd ff:ff:ff:ff:ff:ff
> 3: eth1: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc fq_codel state UP qlen 1000
>     link/ether b8:69:f4:0d:1a:37 brd ff:ff:ff:ff:ff:ff
>     inet6 fe80::ba69:f4ff:fe0d:1a37/64 scope link 
>        valid_lft forever preferred_lft forever

> current ip a
> 2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc fq_codel state UP qlen 1000
>     link/ether e2:3e:65:92:0b:bb brd ff:ff:ff:ff:ff:ff

That's wrong. This is a random mac.

> 3: eth1: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc fq_codel state DOWN qlen 1000
>     link/ether b8:69:f4:0d:1a:36 brd ff:ff:ff:ff:ff:ff
> ```

I'll review the patchset when it doesn't include bits from #3348 

HTH

f00b4r0 commented:
> > That's wrong. This is a random mac.
> 
> Where can I fix that?

You don't define lan interface in 02_network.

Looking at this comment https://github.com/openwrt/openwrt/pull/3348#discussion_r477189194 from @adschm, I'm going to have to disagree: the `ucidef_set_interfaces_lan_wan` statement _is_ necessary. Otherwise mac setup won't work.

Coincidentally I suspect if you check your hAP AC lite, you will notice the same problem.

f00b4r0 commented:
My previous comment can be ignored. AIUI the invalid MAC address on eth0 is an accepted side effect of the current way we set MAC addresses.

f00b4r0 committed:
ath79: add support for MikroTik RouterBOARD hAP ac lite

The MikroTik RB952Ui-5ac2nD (sold as hAP ac lite) is an indoor 2.4Ghz
and 5GHz AP/router with a 2 dBi integrated antenna.

See https://mikrotik.com/product/RB952Ui-5ac2nD for more details.

Specifications:
 - SoC: QCA9533
 - RAM: 64MB
 - Storage: 16MB NOR
 - Wireless: QCA9533 802.11b/g/n 2x2 / QCA9887 802.11a/n/ac 2x2
 - Ethernet: AR934X switch, 5x 10/100 ports,
    10-28 V passive PoE in port 1, 500 mA PoE out on port 5
 - 6 user-controllable LEDs:
   - 1x user (green)
   - 5x port status (green)

Flashing:
 TFTP boot initramfs image and then perform sysupgrade. The "Internet"
 port (port number 1) must be used to upload the TFTP image, then
 connect to any other port to access the OpenWRT system.
 Follow common MikroTik procedure as in
 https://openwrt.org/toh/mikrotik/common.

Signed-off-by: Thibaut VARÈNE <hacks@slashdirt.org>

f00b4r0 commented:
@mkg20001 OK so I eventually rewrote the hAP ac lite PR from scratch (see #9825), and I noticed that there were quite a lot more that was wrong than what I initially spotted in the original PR. I would suggest you take a look to adjust your own PR (in particular there's a lot of redundant bits in your DTS that can be removed). HTH.

mkg20001 commented:
It's not booting. What's the easiest way to get serial? Preferably without taking anything apart.

f00b4r0 commented:
by the way, the hAP doesn't have PCI, so please don't blindly copy everything.

mkg20001 commented:
I still have the random mac issue.
```
2: eth0: <BROADCAST,MULTICAST> mtu 1500 qdisc fq_codel state DOWN qlen 1000
    link/ether ae:4d:b3:51:4e:4a brd ff:ff:ff:ff:ff:ff
3: eth1: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc fq_codel master br-wan state UP qlen 1000
    link/ether 32:de:96:ad:1a:77 brd ff:ff:ff:ff:ff:ff
```

@f00b4r0 does it occur on your device or is it just the hAP that's borked?

f00b4r0 commented:
> I still have the random mac issue.
> 
> ```
> 2: eth0: <BROADCAST,MULTICAST> mtu 1500 qdisc fq_codel state DOWN qlen 1000
>     link/ether ae:4d:b3:51:4e:4a brd ff:ff:ff:ff:ff:ff
> 3: eth1: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc fq_codel master br-wan state UP qlen 1000
>     link/ether 32:de:96:ad:1a:77 brd ff:ff:ff:ff:ff:ff
> ```
> 
> @f00b4r0 does it occur on your device or is it just the hAP that's borked?

the eth* MACs are irrelevant. Look at br-lan and wan interfaces MACs. Check the content of `/etc/board.json` you should have the correct MACs there. Otherwise flash clean without preserving settings.

f00b4r0 commented:
Please make sure to fix the alphabetical order of the makefile and board.d/ entries when you'll next rebase. 951 comes before 952 ;)

f00b4r0 commented on target/linux/ath79/mikrotik/base-files/etc/board.d/02_network:
```diff
@@ -24,6 +24,12 @@ ath79_setup_interfaces()
 	mikrotik,routerboard-wapr-2nd)
 		ucidef_set_interface_lan "eth0"
 		;;
+	mikrotik,routerboard-952ui-5ac2nd|\
+	mikrotik,routerboard-951ui-2nd)
```
missed one :)

mkg20001 commented on target/linux/ath79/mikrotik/base-files/etc/board.d/02_network:
```diff
@@ -24,6 +24,12 @@ ath79_setup_interfaces()
 		;;
+	mikrotik,routerboard-951ui-2nd|\
+	mikrotik,routerboard-952ui-5ac2nd)
+		ucidef_set_interfaces_lan_wan "eth0" "eth1"
```
is this correct? I changed it to include eth1 as wan

f00b4r0 replied:
No. You shouldn't need to change anything from my PR. The hAP is identical to hAP ac lite minus the PCI 802.11ac wireless.

mkg20001 committed:
ath79: add support for MikroTik hAP (RB951Ui-2nD)

The MikroTik hAP (product code RB951Ui-2nD) is
an indoor 2.4Ghz AP with a 2 dBi integrated antenna built around the
Atheros QCA9531 SoC.

Specifications:
 - SoC: Atheros QCA9531
 - RAM: 64 MB
 - Storage: 16 MB NOR - Winbond 25Q128FVSG
 - Wireless: Atheros QCA9530 (SoC) 802.11b/g/n 2x2
 - Ethernet: Atheros AR934X switch, 5x 10/100 ports,
   10-28 V passive PoE in port 1, 500 mA PoE out on port 5
 - 8 user-controllable LEDs:
   · 1x power (green)
   · 1x user (green)
   · 4x LAN status (green)
   · 1x WAN status (green)
   · 1x PoE power status (red)

See https://mikrotik.com/product/RB951Ui-2nD for more details.

Notes:
 The device was already supported in the ar71xx target.

Flashing:
 TFTP boot initramfs image and then perform sysupgrade. Follow common
 MikroTik procedure as in https://openwrt.org/toh/mikrotik/common.

Signed-off-by: Maciej Krüger <mkg20001@gmail.com>

mkg20001 commented:
merged in 5ce64e0646

kmarzic commented:
Hi,

I have [RB951G-2HnD](https://mikrotik.com/product/RB951G-2HnD), which is [very similar](https://mikrotik.com/products/compare/RB951G-2HnD+RB951Ui-2HnD) to [RB951ui](https://mikrotik.com/product/RB951Ui-2HnD).

On RB951G-2HnD I am running 19.07.10, which is latest supported version.

How hard (or is it possible) to reuse code/build for RB951ui and use it on RB951ui to run 22.03?

Thanks!

mkg20001 commented:
@kmarzic You can use `#include "qca9533_mikrotik_routerboard-95x.dtsi"` in your DTS file, it should be a base for all boards of that type.

Take a look at my commit, more or less the same for your device (except the additional hardware is different): [`625434c` (#4967)](https://github.com/openwrt/openwrt/pull/4967/commits/625434c55e71388a423f57fb5274fce594eb8dda)

f00b4r0 commented:
> Hi,
> 
> 
> 
> I have [RB951G-2HnD](https://mikrotik.com/product/RB951G-2HnD), which is [very similar](https://mikrotik.com/products/compare/RB951G-2HnD+RB951Ui-2HnD) to [RB951ui](https://mikrotik.com/product/RB951Ui-2HnD).
> 
These are NAND devices. OpenWrt no longer supports mikrotik NAND devices.
